src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+                    if bExistHandleCollection then 
+                        handleCollection.TryGetValue(jobObject)
+                    else
+                        false, Unchecked.defaultof<_>
+                let ticksRef = 
+                    if not bExist then 
+                        ref DateTime.MinValue.Ticks // Continue to execute 
+                    else
+                        let _, _, ticksRef = tuple
+                        ticksRef
+                let bCancelled = ((!ticksRef) = DateTime.MaxValue.Ticks)
+                if not bCancelled then 
+                    ticksRef := DateTime.UtcNow.Ticks
+                    continuation() 
+                    if Utils.IsNotNull unblockHandle  then 
+                        // If there is an unblock handle, set it. 
+                        unblockHandle.Set() |> ignore  
+                if bExist then 
+                    let rwh, _, _ = tuple 
+                    rwh.Unregister(null) |> ignore 
+                    handleCollection.TryRemove( jobObject ) |> ignore 
+                    if handleCollection.IsEmpty then
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle ) |> ignore 
+                    ThreadPoolWait.ToUnregister()
+                else
+                    rwhToRemove.Enqueue( handle, jobObject )
+            with 
+            | ex -> 
+                Logger.LogF( LogLevel.Error, fun _ -> sprintf "ThreadPoolWait.Callback, exception: %A" ex )
+                reraise()
+        else
+            failwith "ThreadPoolWait.Callback, timeout path reached (impossible execution route)"
+    static member ToUnregister() = 
+        let requeue = List<_>()        
+        while not rwhToRemove.IsEmpty do
+            let bRemove, removeTuple = rwhToRemove.TryDequeue()
+            if bRemove then 
+                let handle, jobObject = removeTuple
+                let bExistHandleCollection, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+                if bExistHandleCollection then 
+                    let bExist, tuple = handleCollection.TryGetValue(jobObject)  
+                    if bExist then 
+                        let rwh, _, _ = tuple       
+                        rwh.Unregister(null) |> ignore 
+        ()
+    static member TryTerminateAll() = 
+        for pair0 in ThreadPoolWait.WaitHandleCollection do
+            ThreadPoolWait.TryRemove( pair0.Key ) |> ignore 
+        ()
+    /// Mark all functions associated with handle as cancelled, the continuation function will not execute
+    static member TryRemove (handle:WaitHandle) = 
+        let bExist, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+        if bExist then 
+            for pair in handleCollection do 
+                let rwh, _, ticksRef = pair.Value
+                // Try to marked job as cancelled, 
+                Volatile.Write( ticksRef, DateTime.MaxValue.Ticks )
```

Comment:
Just to call out that this is another place that we rely on 64-bit read/write being atomic.